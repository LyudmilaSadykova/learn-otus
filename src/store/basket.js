import { defineStore } from "pinia";
import { ref } from "vue";

export const useBasketStore = defineStore('basket', () => {
    const basketList = ref([])

    function addToBasket(product) {
        basketList.value.push(product);
    }

    function clearBasket() {
        basketList.value = [];
    }

    function delToBasket(id) {
        basketList.value = JSON.parse(JSON.stringify(basketList.value.filter((item) => item.id !== id)));
    }

    return { basketList, addToBasket, clearBasket, delToBasket }
})