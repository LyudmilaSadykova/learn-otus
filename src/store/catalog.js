import { defineStore } from "pinia";
import { ref, computed } from "vue";
import axios from "axios";

export const useCatalogStore = defineStore('catalog', () => {
    const loading = ref(false);
    const productList = ref([])
    const filters = ref([])

    async function loadProducts() {
        loading.value = true;

        await axios.get('https://fakestoreapi.com/products')
        .then(response => {
            productList.value = response.data
            //console.log(productList.value);
        })
        .catch(error => {
            console.log(error);
        })
        .finally(() => loading.value = false)
    }

    const productListHome = computed(() => [...productList.value].splice(0, 4))

    function updateFilters(filterList) {
        filters.value = filterList;
    }

    const productListFiltered = computed(() => {
        return [...productList.value].filter((product) => 
            ((!filters.value[0] || filters.value[0] && product.title.indexOf(filters.value[0]) > -1) &&
            (!filters.value[1] || filters.value[1] && product.price > filters.value[1]) && (!filters.value[2] || filters.value[2] && product.price < filters.value[2]))
          );
    })

    function addProduct(product) {
        productList.value.push(product);
    }

    return { loading, productList, loadProducts, productListHome, updateFilters, productListFiltered, addProduct }
})