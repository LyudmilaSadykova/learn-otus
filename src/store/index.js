import { createStore } from 'vuex';
import axios from 'axios';

export default createStore({
    state: {
        productList: [],
        filters: [],
        basketList: []
    },
    getters: {
        getProducts(state) {
          return state.productList;
        },
        filteredProducts(state) {
            return [...state.productList].filter((product) => 
                ((!state.filters[0] || state.filters[0] && product.title.indexOf(state.filters[0]) > -1) &&
                (!state.filters[1] || state.filters[1] && product.price > state.filters[1]) && (!state.filters[2] || state.filters[2] && product.price < state.filters[2]))
              );
        },
        getBasketList(state) {
            return state.basketList;
        }
    },
    mutations: {
        SET_PRODUCTS(state, products) {
            state.productList = products;
        },
        PUSH_PRODUCT(state, product) {
            state.productList.push(product);
        },
        SET_FILTERS(state, filterList) {
            state.filters = filterList;
        },
        PUSH_BASKET(state, product) {
            state.basketList.push(product);
        },
        SET_BASKET(state) {
            state.basketList = [];
        },
    },
    actions: {
        /*async loadProducts({ commit }) {
            await axios.get('https://fakestoreapi.com/products')
            .then(response => {
                commit('SET_PRODUCTS', response.data)
                //console.log(productList.value);
            })
            .catch(error => {
                console.log(error);
            })
        },*/
        updateProducts({ commit }, productList) {
            commit('SET_PRODUCTS', productList)
        },
        addProduct({ commit }, product) {
            commit('PUSH_PRODUCT', product)
        },
        updateFilters({ commit }, filterList) {
            commit('SET_FILTERS', filterList)
        },
        addToBasket({ commit }, product) {
            commit('PUSH_BASKET', product)
        },
        clearBasket({ commit }) {
            commit('SET_BASKET')
        },
    }
    
})