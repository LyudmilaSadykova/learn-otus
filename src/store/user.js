import { defineStore } from "pinia";
import { ref } from "vue";

export const useAccountStore = defineStore('account', () => {
    const accountData = ref([])
    const isAuth = ref(false)

    function saveData(data) {
        accountData.value = data;
    }

    function loginUser(token) {
        localStorage.setItem('token', token);
        isAuth.value = true;
    }

    function logoutUser() {
        localStorage.removeItem('token');
        isAuth.value = false;
    }

    return { accountData, isAuth, saveData, loginUser, logoutUser }
})