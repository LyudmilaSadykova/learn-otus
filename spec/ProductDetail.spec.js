import { mount } from '@vue/test-utils';
import component from '../src/components/ProductDetail.vue';
import { createTestingPinia } from '@pinia/testing';
import { useBasketStore } from '../src/store/basket';
import { useRoute } from 'vue-router';
import { expect, vi } from 'vitest';
import axios from 'axios';

vi.mock('vue-router');

const mockRoute = useRoute.mockImplementation(() => ({
    params: {
        productId: '1',
    },
}));

function getProductList(id) {
    return new Promise((resolve, reject) => {
        axios.get('https://fakestoreapi.com/products/' + id).then((result) => {
            setTimeout(resolve(result.data), 15000);
        }).catch((err) => {
            reject(err);
        })
    })
 }

 const product = {"id":1,"title":"Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops","price":109.95,"description":"Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday","category":"men's clothing","image":"https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg","rating":{"rate":3.9,"count":120}};

describe('ProductDetail component', () => {
    const buildWrapper = () => mount(component, {
        global: {
            plugins: [createTestingPinia({
                createSpy: vi.fn,
                stubActions: false,
                initialState: {
                    basketList: []
                }}
            )],
        },
    });

    it('mount without errors', () => {
        const wrapper = buildWrapper();
        expect(wrapper.exists()).toBe(true);
        expect(wrapper.element).toMatchSnapshot();
    });

    it('function exist', () => {
        expect(typeof getProductList).toBe('function');
      });

    it('load product sussess', async () => {
        return await expect(getProductList(1))
        .resolves
        .toStrictEqual({"id":1,"title":"Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops","price":109.95,"description":"Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday","category":"men's clothing","image":"https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg","rating":{"rate":3.9,"count":120}})
    }, 15000);

    it('load product error', async () => {
        return await expect(getProductList(0))
        .resolves
        .toStrictEqual("")
    }, 15000);

    it('emit event add to cart', async () => {
        const wrapper = buildWrapper();

        const basket = useBasketStore();
        const button = wrapper.find('v-btn');
        expect(button.exists()).toBe(true);
        // await button.trigger('click');

        basket.addToBasket(product)
        basket.addToBasket(product)
        expect(basket.addToBasket).toHaveBeenCalledTimes(2);
        expect(basket.basketList.length).equal(2)
    });

});
