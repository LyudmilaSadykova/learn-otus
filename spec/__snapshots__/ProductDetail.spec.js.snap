// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductDetail component > mount without errors 1`] = `
<div
  class="detail"
  data-v-4e42be35=""
>
  <v-card
    class="mx-auto"
    color="indigo-darken-3"
    data-v-4e42be35=""
    hover=""
    variant="text"
  >
    <v-card-title
      data-v-4e42be35=""
    >
      
    </v-card-title>
    <div
      class="d-flex flex-no-wrap justify-space-between"
      data-v-4e42be35=""
    >
      <div
        data-v-4e42be35=""
      >
        <v-card-subtitle
          data-v-4e42be35=""
        >
          
        </v-card-subtitle>
        <v-card-text
          data-v-4e42be35=""
        >
          
        </v-card-text>
        <v-card-text
          data-v-4e42be35=""
        >
           Price: $ 
        </v-card-text>
        <!--v-if-->
        <v-card-actions
          data-v-4e42be35=""
        >
          <v-btn
            data-v-4e42be35=""
            variant="tonal"
          >
             Добавить в корзину 
          </v-btn>
        </v-card-actions>
      </div>
      <v-avatar
        class="ma-3"
        data-v-4e42be35=""
        rounded="0"
        size="200"
      >
        <div
          class="box"
          data-v-4e42be35=""
        />
      </v-avatar>
    </div>
  </v-card>
</div>
`;
